pkg/exporter/ipfix_proto.go

```diff
@@ -0,0 +1,498 @@
+package exporter
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+
+	"github.com/netobserv/netobserv-ebpf-agent/pkg/flow"
+	"github.com/sirupsen/logrus"
+	"github.com/vmware/go-ipfix/pkg/entities"
+	ipfixExporter "github.com/vmware/go-ipfix/pkg/exporter"
+	"github.com/vmware/go-ipfix/pkg/registry"
+)
+
+var ilog = logrus.WithField("component", "exporter/IPFIXProto")
+
+type IPFIXProto struct {
+	hostPort     string
+	exporter     *ipfixExporter.ExportingProcess
+	templateIDv4 uint16
+	templateIDv6 uint16
+}
+
+func addElementToTemplate(log *logrus.Entry, elementName string, value []byte, elements *[]entities.InfoElementWithValue) error {
+	element, err := registry.GetInfoElement(elementName, registry.IANAEnterpriseID)
+	if err != nil {
+		log.WithError(err).Errorf("Did not find the element with name %s", elementName)
+		return err
+	}
+	ie, err := entities.DecodeAndCreateInfoElementWithValue(element, value)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to decode element %s", elementName)
+		return err
+	}
+	*elements = append(*elements, ie)
+	return nil
+}
+
+func makeByteFromUint8(value uint8) []byte {
+	bs := make([]byte, 1)
+	bs[0] = value
+	return bs
+}
+
+func makeByteFromUint16(value uint16) []byte {
+	bs := make([]byte, 2)
+	binary.BigEndian.PutUint16(bs, value)
+	return bs
+}
+
+func makeByteFromUint32(value uint32) []byte {
+	bs := make([]byte, 4)
+	binary.BigEndian.PutUint32(bs, value)
+	return bs
+}
+
+func makeByteFromUint64(value uint64) []byte {
+	bs := make([]byte, 8)
+	binary.BigEndian.PutUint64(bs, value)
+	return bs
+}
+
+// Flow Record to corresponding IPFIX Types
+// EthProtocol -> ethernetType
+// Direction   -> flowDirection
+// SrcMac      -> sourceMacAddress
+// DstMac      -> destinationMacAddress
+// SrcAddr(v4) -> sourceIPv4Address
+// DstAddr(v4) -> destinationIPv4Address
+// SrcAddr(v6) -> sourceIPv6Address
+// DstAddr(v6) -> destinationIPv6Address
+// Protocol    -> protocolIdentifier/nextHeaderIPv6
+// SrcPort     -> sourceTransportPort
+// DstPort     -> destinationTransportPort
+// Bytes       -> octetDeltaCount
+// TimeFlowStart -> flowStartSeconds, flowStartNanoseconds
+// TimeFlowEnd -> flowEndSeconds, flowEndNanoseconds
+// Packets     -> packetDeltaCount
+// Interface   -> interfaceName
+
+func sendDataRecordv6(log *logrus.Entry, record *flow.Record, exporter *ipfixExporter.ExportingProcess, templateID uint16) error {
+	dataSet := entities.NewSet(false)
+	err := dataSet.PrepareSet(entities.Data, templateID)
+	if err != nil {
+		log.Errorf("Failed in PrepareSet")
+		return err
+	}
+	elements := make([]entities.InfoElementWithValue, 0)
+
+	err = addElementToTemplate(log, "ethernetType", makeByteFromUint16(record.EthProtocol), &elements)
```

Comment:
Is the IPFIX exporter performing correctly? It looks to me that invoking `addElementToTemplate` for each record, for each field, is something that can be optimized.

Basically because `addElementToTemplate` invokes `registry.GetInfoElement` (a lookup) and a decode... I wonder whether this can be cached somewhere.

Reply:
Yes, you are right, I will store the set in the global struct, so that we do not need to create it again for each record.

Reply:
I have cached the ieValues now.